### PR TITLE
Add API requirements and design document

### PR DESCRIPTION
**Pull request by matthiasbeyer**
As discussed in our call today, this patch adds the document we crafted.

@didier-wenzek and @makr11st I put your "Signed-off-by" into the commit message as well, as we all collaborated here. I hope that's okay with you.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [x] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

File: design/API_requirements_and_design.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  - The motivation is to be able to deliver a battery included executable, ready
+    to be used for a large set of use-cases without any Rust expertise.
+  - Notably a battery-included executable should include an MQTT connection
+    plugin and a JSON plugin for telemetry messages.
+- Plugins are instantiated and configured through an external configuration at
+  runtime
+- At startup the configuration is read and from it a set of plugins 'to be
+  instantiated' is derived
+- One should be able to deduplicate parts of the configuration (to prevent
+  repetition)
```
Comment (didier-wenzek) [inline design/API_requirements_and_design.md:34]: I'm not expecting something fancy here - just a way to share values between instances of the same plugin.

Looking at this [configuration example](https://github.com/matthiasbeyer/thin-edge.io/blob/feature/add_tedge_api_impl_typed/tedge/example-config.toml),
I wonder if this can be simply done by adding a level for all the instances of a plugin.

For instance:
```
[plugins.mqtt.configuration]
host = "127.0.0.1"
port = 1883
clear_session = false

[plugins.mqtt.c8y.configuration]
session = "c8y-mapper"
subscriptions = "c8y/#"

[plugins.mqtt.tedge.configuration]
session = "tedge"
subscriptions = "tedge/measurements#"
```

File: design/API_requirements_and_design.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+- Should plugin instances be changeable during runtime? (open question)
+- Should some plugin configurations be changeable at runtime or not?
+- Having an optional plugin handling SIGHUP and send an internal message
+  UpdateConfig to all the running plugin instances.
+- Would having multiple configuration files be an option? (open question)
```
Comment (didier-wenzek) [inline design/API_requirements_and_design.md:39]: I think we need to put apart two aspects and to make this separation explicit with two kind of configuration files:

* Plugin instantiation and inter-connection. This defines the application with its set of features.
* Runtime configuration (IP addresses, external resources, internal capacity tuning ...). This defines the deployment of the application in a given operational context.

One should be able to update one aspect without risks to break the other. Furthermore, these two aspects are defined by different persons with different schedules.

From my point of view, a configuration akin to [this example](https://github.com/matthiasbeyer/thin-edge.io/blob/feature/add_tedge_api_impl_typed/tedge/example-config.toml) mixes the two aspects and will be hard to maintain for a realistic application.

File: design/API_requirements_and_design.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+- If a plugin does not support a message that another plugin expects it to, the
+  runtime should not start
+  - Same thing for a missing required peer for a given kind, or two peers when
+    just one is expected.
+- Plugin/Application shutdown
+- Shutdown is signalled to all plugins, giving them a possibility to handle this
+  case
+- A plugin to handle SIGTERM? Or part of the core?
+- Plugins should be free to communicate with any other plugins, even in a
+  circular fashion.
```
Comment (TheNeikos) [inline design/API_requirements_and_design.md:55]: One aspect I am missing, and just now remembered is that communication between plugins should be fixed by the configuration. This makes the relationship between plugins explicit to the user and ultimately easier to use.

**Review by TheNeikos — CHANGES_REQUESTED** [review]

File: design/API_requirements_and_design.md (hunk 3 above)
Comment (TheNeikos) [reply design/API_requirements_and_design.md:55]: To expand a bit:

Users do not necessarily get to choose which features a certain TE application has included. Having any form of 'implicit' communication (by for example sending it to the next best plugin that happens to accept the message type) makes using the application very unintuitive. 

Some example situations where this can happen:

 - Two plugins that happen to both respond to some SoftwareManagement messages but handle different overlapping-subsets of it. If you were to send a message that both are handling, you'd get duplicate messages and potentially duplicate actions which can be dangerous.
 - Enabling a plugin without keeping in mind _everything else_ in the configuration can potentially start causing effects on the device that the user did not intend to make, simply because now a target 'became available'

Comment (matthiasbeyer) [reply design/API_requirements_and_design.md:55]: >> Enabling a plugin without keeping in mind everything else in the configuration can potentially start causing effects on the >>  device that the user did not intend to make, simply because now a target 'became available'
>
> I don't understand this point.

I guess something like this: Plugins A, B and C are currently enabled and running. They work fine and properly. If now a plugin D gets enabled, and thin-edge now "figures out" a new way how these four plugins "should" communicate, simply enabling D might break the previously working setup of A, B and C.

That's why @TheNeikos proposes:

> [...] communication between plugins should be fixed by the configuration. This makes the relationship between plugins explicit to the user and ultimately easier to use.

---

On the other part of your comment:

> Some plugins might need to dispatch messages among a set of plugins delivering the same service but at different level/scope/target, this dispatch being done after a key extracted from the messages.

This sounds like the plugin that dispatches the messages already knows _when_ to dispatch these (lets call them) "special" messages to another plugin, right?
That would mean that the plugin would simply implement this logic: "If the message is special, send to plugin A, else to plugin B".

---

> On the other hand, other connections define the application and it makes little sense to change this connection in operations. For example a c8yplugin is expecting an mqtt_connection to c8y and a software_management_plugin.

I'm not entirely sure I understand your point here, could you maybe rephrase it?

File: design/API_requirements_and_design.md (hunk 2 above)
Comment (makr11st) [reply design/API_requirements_and_design.md:39]: I suppose one way to get out of this would be to have generic configuration (as it is proposed now) but allow for inclusion of multiple files, this could be achieved by adding a clause in config `include` which would be a path allowing for pointing to a file or directory and leave to the user to decide if they prefer to have everything in a single file or multiple ones. 
The configuration reading phase would then collect all the configs, validate them and load.

File: design/API_requirements_and_design.md (hunk 3 above)
Comment (didier-wenzek) [reply design/API_requirements_and_design.md:55]: > I guess something like this: Plugins A, B and C are currently enabled and running. They work fine and properly. If now a plugin D gets enabled, and thin-edge now "figures out" a new way how these four plugins "should" communicate, simply enabling D might break the previously working setup of A, B and C.
That's why @TheNeikos proposes:
>> [...] communication between plugins should be fixed by the configuration. This makes the relationship between plugins explicit to the user and ultimately easier to use.

Okay, I got it. I fully agree that the system must not try to be smart (e.g. connecting plugins because the message types match).

However, I'm still not sure that this should be configured *only* at runtime (in some `config.toml`). I see cases where compile time configuration (in `main.rs`) would be more appropriate because there is no other meaningful alternatives.

I tried to give an example here:

>>On the other hand, other connections define the application and it makes little sense to change this connection in operations. For example a c8yplugin is expecting an mqtt_connection to c8y and a software_management_plugin.

> I'm not entirely sure I understand your point here, could you maybe rephrase it?

But was not clear enough. I think that some plugin connections, as between a `c8yplugin` and its `mqtt_connection` to c8y, there is no point to say again and again, in every runtime configuration and process launches, that this `c8y_mqtt_connection` instance is the mqtt connection of the `c8yplugin` instance.

Comment (didier-wenzek) [reply design/API_requirements_and_design.md:55]: >> Some plugins might need to dispatch messages among a set of plugins delivering the same service but at different level/scope/target, this dispatch being done after a key extracted from the messages.

> This sounds like the plugin that dispatches the messages already knows when to dispatch these (lets call them) "special" messages to another plugin, right?
That would mean that the plugin would simply implement this logic: "If the message is special, send to plugin A, else to plugin B".

Definitely, the dispatcher plugin must implement the dispatching logic. However, this logic must be compatible with the actual plugin connections - especially when these ones are configured at runtime using plugin external names. Is the plugin named "A" in the config the correct target for the A messages?

Comment (matthiasbeyer) [reply design/API_requirements_and_design.md:55]: > However, I'm still not sure that this should be configured only at runtime (in some config.toml). I see cases where compile time configuration (in main.rs) would be more appropriate because there is no other meaningful alternatives.

Ah, I guess now I understand. You mean that someone might implement a usecase in more than one plugin, say (for the sake of an example), they implement their usecase in 3 plugins: A, B, C. One plugin (A) that actually talks to other plugins plus two "backend" plugins (B, C), that are only ever plugged into A. In this case, it would only make sense to provide _configuration_ for B and C in the `config.toml`, but never "wiring".

Does this represent what you mean?

Comment (didier-wenzek) [reply design/API_requirements_and_design.md:55]: >> However, I'm still not sure that this should be configured only at runtime (in some config.toml). I see cases where compile time configuration (in main.rs) would be more appropriate because there is no other meaningful alternatives.

> Ah, I guess now I understand. You mean that someone might implement a usecase in more than one plugin, say (for the sake of an example), they implement their usecase in 3 plugins: A, B, C. One plugin (A) that actually talks to other plugins plus two "backend" plugins (B, C), that are only ever plugged into A. In this case, it would only make sense to provide configuration for B and C in the config.toml, but never "wiring".
Does this represent what you mean?

Yes. Though, said like this is sounds that it would be simpler to have a huge A-B-C plugin including the features of the 3 former ones.

Comment (matthiasbeyer) [reply design/API_requirements_and_design.md:55]: Exactly that would have been my next question :laughing: :+1:

File: design/API_requirements_and_design.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+* All these tedge plugins crates depends on the tedge_api crate that defines how
+  to create, connect and run plugin instances.
+* Each tedge plugin crate can define its own types of messages. For instance, a
+  tedge_telemetry plugin can defines measurements, events, alarms and setpoints
+  messages.
+* To exchange typed messages, plugin crates need dependency relationship. For
+  instance, the tedge_json_over_mqtt plugin crate will have to depend on the
+  tedge_telemetry plugin crate and the tedge_mqtt crate.
+* The tedge_api crate defines only messages related to the runtime as Shutdown.
```
Comment (matthiasbeyer) [inline design/API_requirements_and_design.md:13]:
```suggestion
* The tedge_api crate defines only messages related to the runtime as as `Shutdown`.
```

File: design/API_requirements_and_design.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+* The implementation of the mechanisms defined by the tedge_api are provided by
+  a tedge_runtime crate.
+* The final executable is build as an assemblage of all these plugin crates and
+  a `main.rs` that defines which plugins have to be instantiated and then
+  launched.
```
Comment (matthiasbeyer) [inline design/API_requirements_and_design.md:18]:
```suggestion
  a `main.rs` that defines which plugins can potentially be instantiated and
  launched.
```

File: design/API_requirements_and_design.md [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+One should be able to build a thin-edge executable from tedge plugins crates that have been implemented independently:
+
+* All these tedge plugins crates depends on the tedge_api crate that defines how
+  to create, connect and run plugin instances.
+* Each tedge plugin crate can define its own types of messages. For instance, a
+  tedge_telemetry plugin can defines measurements, events, alarms and setpoints
+  messages.
+* To exchange typed messages, plugin crates need dependency relationship. For
+  instance, the tedge_json_over_mqtt plugin crate will have to depend on the
+  tedge_telemetry plugin crate and the tedge_mqtt crate.
```
Comment (matthiasbeyer) [inline design/API_requirements_and_design.md:12]: I am not sure we should even write this in the document. Because to exchange typed messages, plugins only need to know the type, but not whom it may come from.

Think of this:

* A crate `telemetry_types` defines `Measurement`
* A plugin `plugin_c8y` can send `Measurement`s
* A plugin `plugin_measurement_to_json` can receive `Measurement`s, turn it into `Vec<u8>` and send that
* A plugin `plugin_mqtt` can receive `Vec<u8>` and send that to an MQTT broker

None of the above plugins actually need to know that the other plugins actually exist! Only `plugin_c8y` and `plugin_measurement_to_json` need to depend on `telemetry_types`, but not even on eachother!

So maybe we should rephrase that paragraph, or leave it out if we cannot think of something else?

File: design/API_requirements_and_design.md [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+  best of its abilities
+- The runtime should not be able to start with a clear broken configuration
+- One plugin panicking should not impact other plugins or the runtime
+- If a plugin does not support a message that another plugin expects it to, the
+  runtime should not start
+  - Same thing for a missing required peer for a given kind, or two peers when
+    just one is expected.
+- Plugin/Application shutdown
+- Shutdown is signalled to all plugins, giving them a possibility to handle this
+  case
```
Comment (matthiasbeyer) [inline design/API_requirements_and_design.md:52]:
```suggestion
- Plugin/Application shutdown
  - Shutdown is signalled to all plugins, giving them a possibility to handle this
  case
```

I believe that the first line should be a heading for the sub-points, right? :laughing:

**Review by matthiasbeyer — COMMENTED** [review]
While reviewing this document, I found some more points we might need to talk or think about.

I guess these are rather minor.

File: design/API_requirements_and_design.md (hunk 4 above)
Comment (didier-wenzek) [reply design/API_requirements_and_design.md:13]: Not sure to grasp the point.

File: design/API_requirements_and_design.md (hunk 5 above)
Comment (didier-wenzek) [reply design/API_requirements_and_design.md:18]: This is something we have indeed to discuss. Somehow I see this change (can be vs have to) as a great source of complexity.

* This is why I wrote "have to" and work in that direction with [my proposal](https://github.com/thin-edge/thin-edge.io/pull/1153).
* But I also appreciate the benefits of the more complex but flexible [approach you proposed with Marcel](https://github.com/thin-edge/thin-edge.io/pull/979)

I open to discuss this point further.

File: design/API_requirements_and_design.md (hunk 6 above)
Comment (didier-wenzek) [reply design/API_requirements_and_design.md:12]: We are fully aligned!

The sentence might be rewritten using the details you gave in your comment.

File: design/API_requirements_and_design.md (hunk 4 above)
Comment (matthiasbeyer) [reply design/API_requirements_and_design.md:13]: Hm, somehow this suggestion got messed up. Sorry.

File: design/API_requirements_and_design.md (hunk 6 above)
Comment (matthiasbeyer) [reply design/API_requirements_and_design.md]: Added a fixup commit to address this, but feel free to suggest more, I kept the point rather short.

File: design/API_requirements_and_design.md (hunk 5 above)
Comment (TheNeikos) [reply design/API_requirements_and_design.md:20]: Complexity for whom? As I believe that to be the most critical aspect of having such an abstraction.

To help the discussion here's some definitions of different people that might use TE:

- Framework Developers (us)
- Framework Users (people that develop plugins/custom deployment of TE)
- TE Users (people that use a built binary and have to configure/deploy it)
- End Users (people that interface with a product/deployment of TE)

Here is what each person gains/loses when supporting optional instantiation of components from my point of view:

- End Users do not see any direct change that is quantifiable
- TE Users get more flexibility on their deployments as a single compiled binary can be re-purposed. It would be seen as a swiss-army knife of edge configuration. Losses are slightly larger binary sizes; more complexity to understand what component does what and how to configure them.
- Framework Users do not see any direct change as writing a component will still be done through the interface of traits in some fashion.
- Framework Developers is once again hard to quantify, but in general things use more advanced rust concepts like trait objects and dynamic dispatch.


And here is what each person were to gain/lose when having fixed instantiation of components:

- End Users do not see any direct change that is quantifiable
- TE Users only have a fixed configuration to fill out, creating a new deployment requires compilation of a new binary. Losses are needing the infrastructure/knowledge to create the new binaries. Updating the behaviour of a deployment gets harder as more than just the configuration needs to change.
- Framework Users do not see any direct change as writing a component will still be done through the interface of traits in some fashion.
- Framework Developers is hard to quantify due to the question of how abstract the innards are designed.



So overall, I believe that being optional on components is a win for TE users as they get flexibility at low cost. This does make it more complex for framework developers (us), but this is all a finite amount of effort as the Rust concepts are learnable (and teachable).

Comment (didier-wenzek) [reply design/API_requirements_and_design.md:20]: @TheNeikos this is a fair and helpful comparison.

Comment (didier-wenzek) [reply design/API_requirements_and_design.md:20]: I changed my mind on this point. I agree that it will convenient for the agent developers to be able to configure at runtime which components are instantiated and launched.

**Comment by didier-wenzek** [comment]
Superseded by https://github.com/thin-edge/thin-edge.io/pull/1549
